### PR TITLE
Allow universe annotations in record projections.

### LINKED ISSUE #11054: Universe annotations in record projections.
Why is it allowed to add explicit universe annotations to a qualified identifier playing the role of a term (e.g. a function), but not to a qualified identifier playing the role of a record field? I.e.:

```
Polymorphic Record foo := { bar : Type }.
Check (bar@{ Type } {| bar := Set |}). (* works fine *)
Check ({| bar := Set |}.(bar@{ Type })). (* does not parse *)
```

Is it an accidental limitation? Would it be useful to have?
If yes, the patch to fix this is trivial: https://github.com/Zimmi48/coq/commit/f6a1ae23621cda29d238c037422d0ca6fde920ef
In particular, it would bring more consistency to the grammar.

### PR DESCRIPTION
**Pull request by Zimmi48**
For consistency with the function application syntax.

I'm submitting this patch as a PR because Hugo thinks it makes sense. I can use the example from #11054 in the test-suite and documentation, but maybe someone who is an actual user of universes has a better example to propose.

**Kind:** enhancement.

Fixes / closes #11054

- [ ] Added / updated test-suite

- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [ ] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by ppedrot** [comment]
Shouldn't it be the converse? The closer we make the projection syntax to the application one, the harder it will be to sort out the current "backward compatibility" mess we have with projections. In particular, in the negative setting, universe indices are just like inductive parameters, and thus unneeded in the projection node.

TL;DR: I think this is going the wrong way.

**Comment by SkySkimmer** [comment]
We could also go in the direction of https://en.wikipedia.org/wiki/Uniform_Function_Call_Syntax

**Comment by ppedrot** [comment]
@SkySkimmer great, we were missing OOP in Coq. :man_facepalming:

**Comment by Zimmi48** [comment]
> In particular, in the negative setting, universe indices are just like inductive parameters, and thus unneeded in the projection node.

Thanks, that's the kind of answer that I was looking for, but did not get so far. By "negative setting", you mean primitive projections, right?

Is there a plan to make records primitive by default and to restrict the projection syntax to primitive records?

**Comment by herbelin** [comment]
I'd like to defend that unneeded does not mean useless! The arguments of a projection can be derived from the parameters in the type of the term to project but it might be useful to support having the inference flow going also the other way round and to derive the parameters of the type of the term to project from the arguments of the projection, say as in:
```
Set Primitive Projections.
Set Universe Polymorphism.
Record A@{i} := {T:Type@{i}}.
Check fun a => T@{Set} a. (* i.e. fun a => a.(T@{Set}) *)
```
More generally, I'm unsure that the question has to do with negative vs positive (where I defined "negative" by "defined from projections", i.e. no `match` for the type). The question is related to the internal representation of implicit arguments in general (even if currently `Primitive Projections` is the only case of syntax supporting dropping implicit arguments) and on whether we want to have the internal syntax match the concrete syntax.

For instance, currently, the following is accepted but the information on `nat` is not kept in the internal representation of `witness` and thus reprinted as `_`:
```
Set Primitive Projections.
Record W A := {witness:A}.
Check fun a => witness nat a.
(* fun a : W nat => witness _ a *)
Check fun a => a.(witness nat).
(* fun a : W nat => witness _ a *)
```
What are the plans then? To forbid the syntax `witness nat a` and `a.(witness nat)` and eventually allow only `a.(witness)`? How feasible is it? Would we e.g. need an automatic translator for that?

In any case, thanks in advance to continue feeding the discussion, as we need to make progresses on the primitive projections question.

**Comment by Zimmi48** [comment]
FWIW, I didn't delete the branch in case one wants to reopen the PR.

**Comment by ppedrot** [comment]
@herbelin I am afraid that artificially keeping a zombie PR open is not going to magically make any progress about it. If anything, the current issue is a design decision that has to be taken e.g. during a WG. Judging from the trivial size of the code change, it can even be reimplemented in a few minutes if needed. This is not a pull *request*, rather a *discussion* request. Thus, I am closing this PR, not to clutter the list of open PRs, which is already quite in need of a cleanup.

**Comment by herbelin** [comment]
> This is not a pull request, rather a discussion request. 

I agree with your arguments. Thanks.
